lib/cli/ui/frame.rb

```diff
@@ -4,6 +4,148 @@ module CLI
+          text = suffix
+          width = CLI::UI::ANSI.printing_width(suffix)
+          end_index = termwidth - 2
+          start_index = end_index - width
+
+          # If the prefix and suffix can't both fit, leave out the suffix
+          # completely.
+          if prefix.end_index > start_index
+            text = ''
+            width = 0
+            start_index = end_index
+          end
+
+          Text.new(text, width, start_index, end_index)
+        end
+
+      end
+
+      # Uses cursor save/restore to more accurately render edges to precise
+      # widths, even when we have emoji that print at semi-variable widths.
+      module MacOSTerminalEdgeRenderer
+        extend EdgeRendererComponents
+
+        def self.render(termwidth:, color:, prefix:, suffix:)
+          prefix = calculate_prefix(prefix)
+          suffix = calculate_suffix(suffix, termwidth, prefix)
+
+          o = +''
+          with_hidden_cursor(o) do
+            reset_cursor(o)
+            set_color(o, color)
+            o << CLI::UI::Box::Heavy::HORZ * termwidth # draw a full line
+            o << print_at_x(prefix.start_index, prefix.text)
+            set_color(o, color)
+            o << print_at_x(suffix.start_index, suffix.text)
+            set_color(o, CLI::UI::Color::RESET)
+          end
+          o << "\n"
+        end
+      end
+
+      # Doesn't use cursor save/restore or other semi-exotic ANSI escape
+      # sequences, which aren't supported by all terminals.
+      module BuildkiteEdgeRenderer
+        extend EdgeRendererComponents
+
+        def self.render(termwidth:, color:, prefix:, suffix:)
+          prefix = calculate_prefix(prefix)
+          suffix = calculate_suffix(suffix, termwidth, prefix)
+
+          interstitial_width = termwidth - (2 + prefix.width + suffix.width)
+
+          o = +''
+          o << prefix.text
+
+          set_color(o, color)
+          o << CLI::UI::Box::Heavy::HORZ * interstitial_width
+
+          o << suffix.text
+
+          set_color(o, color)
+          o << CLI::UI::Box::Heavy::HORZ * 2
+
+          set_color(o, CLI::UI::Color::RESET)
+          o << "\n"
+        end
+      end
+
+      module PipeEdgeRenderer
+        extend EdgeRendererComponents
+
+        # color is unused but necessary for the interface.
+        # rubocop:disable Lint/UnusedMethodArgument
+        def self.render(termwidth:, color:, prefix:, suffix:)
+          prefix = calculate_prefix(prefix)
+          suffix = calculate_suffix(suffix, termwidth, prefix)
+
+          interstitial_width = termwidth - (2 + prefix.width + suffix.width)
+
+          o = +''
+          o << prefix.text
+          o << CLI::UI::Box::Heavy::HORZ * interstitial_width
+          o << suffix.text
+          o << CLI::UI::Box::Heavy::HORZ * 2
+          o << "\n"
+        end
+      end
+
+      class << self
+        attr_accessor :edge_renderer
+      end
+
+      def self.default_edge_renderer
+        if ENV.key?('BUILDKITE')
```

Comment:
Instead of looking for BUILDKITE, can you use `infocmp`, looking for `sc` and `rc`?

Reply:
Potentially, yeah. This is something I've struggled a bit with in general—figuring out how much of cli-ui's behaviour we should be basing on terminfo. It gets more accurate, but more complex, and 99% of our users are on macOS where you have a pretty predictable environment.

I think my inclination would be :-1: unless we're going to overhaul cli-ui to source all this stuff from terminfo. That would be a lot of fun to work on and not an objectively terrible idea.

Reply:
I started on this a long time back: https://github.com/Shopify/cli-ui/compare/master...terminfo

Reply:
I like the idea of an overhaul, but I wasn't suggesting using that for just this case - I'm just wondering if there's something a little more portable that we can use as a dumb-it-down signal than `ENV.key?('BUILDKITE')`. Some way of querying "Do we think you'll handle all of our ANSI stuff? yes/no" and if not, do what you're currently doing for buidlkite.

Reply:
Yeah. The reason I went this way is because I fixed the specific issues that we see on BK. I'm not really sure how best to generalize it. Most interactive TTYs will either support this or not advertise themselves as TTYs.